src/Pages/ChangePassword.cshtml.cs

```diff
@@ -66,17 +70,45 @@ public async Task<IActionResult> OnGet()
 
         private async Task SetCurrency()
         {
-            //Some alert could be displayed that failed to get currency list
             var userId = int.Parse(User.FindFirst(ClaimTypes.Name).Value);
             UserModel = await UserDatabase.GetUserAsync(userId);
             CurrencyOptions = new List<string>();
-            try
+            if (Cache.TryGetValue(CacheKeys.CurrencyList, out List<string> cachedCurrencyList))
             {
-                var currencyList = await CurrencyConverter.GetList();
-                foreach (var currency in currencyList)
+                CurrencyOptions.AddRange(cachedCurrencyList);
+                //Check if user's currency is cached
+                if (Cache.TryGetValue(CacheKeys.UserCurrency, out Currency cachedCurrency))
+                {
+                    if (UserModel.Currency.Equals(cachedCurrency.Code)) return;
+                }
+
+                //If we the code gets here, it means that we don't have the the correct user currency cached
+                try
+                {
+                    var userCurrency = await CurrencyConverter.GetCurrency(UserModel.Currency);
+                    var options = CacheKeys.DefaultCurrencyCacheOptions();
+                    Cache.Set(CacheKeys.UserCurrency, userCurrency, options);
+                }
+                catch (Exception e)
                 {
-                    CurrencyOptions.Add(currency.Code);
+                    _logger.LogInformation(e.ToString());
                 }
+                return;
+            }
+
+            try
+            {
+                var currencyList = await CurrencyConverter.GetList();
+                var currencyCodeList = currencyList.Select(currency => currency.Code).ToList();
+
+                var options = CacheKeys.DefaultCurrencyCacheOptions();
+
+                Cache.Set(CacheKeys.CurrencyList, currencyCodeList, options);
+                CurrencyOptions.AddRange(currencyCodeList);
+
+                //Also set user currency 
+                var userCurrency = currencyList.First(x => x.Code == UserModel.Currency);
+                Cache.Set(CacheKeys.UserCurrency, userCurrency, options);
```

Comment:
So this is the same as previous reading of currencies, just it stores data in user cache and reads only if its empty or outdated data?

Reply:
That's pretty much exactly what it does :D